src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs

```diff
@@ -116,7 +118,18 @@ internal TracerProviderBuilder SetResource(Resource resource)
                 throw new ArgumentNullException(nameof(resource));
             }
 
-            this.resource = resource;
+            this.resourceBuilder = new ResourceBuilder().AddResource(resource);
+            return this;
+        }
+
+        /// <summary>
+        /// Registers a callback that will be executed to configure the <see cref="Resource"/> for the <see cref="TracerProvider"/>.
+        /// </summary>
+        /// <param name="configureResource"><see cref="ResourceBuilder"/>.</param>
+        /// <returns>Returns <see cref="TracerProviderBuilder"/> for chaining.</returns>
+        internal TracerProviderBuilder ConfigureResource(Action<ResourceBuilder> configureResource)
+        {
+            this.configureResources = configureResource ?? throw new ArgumentNullException(nameof(configureResource));
```

Comment:
Should we allow cumulative calls to ConfigureResource to add up, instead of allowing just once?

Reply:
I thought about it, but couldn't decide. Anyone feel strongly?

Reply:
How would this work for dependency injection? e.g. #1469

Reply:
@reyang I don't think we need to do anything special. That PR is registering with DI a callback to configure the TracerProviderBuilder, so that callback can excersie the new `ResourceBuilder` API off of the instance it is passed. We could make it more fancy and allow registration of a dedicated callback specifically for `ResourceBuilder` configuration?

Or were you asking how multiple callbacks would be registered? I'm pretty sure you can register multiple impelementations of an interface with IServiceCollection and ask it for an IEnumerable which will give you back everything registered allowing us to loop through them all.